### PR TITLE
SLURM migration for SCXA control workflow

### PR DESCRIPTION
**Pull request by pmb59**
This PR's goal is simply to migrate from LSF to SLURM.

### CODE REVIEW

**Comment by pmb59** [comment]
Some env variables need to be added, for instance to replace 
```
    queue="$SCXA_LSF_QUEUE"
    clusterOptions="$SCXA_LSF_OPTIONS"
```

File: nextflow.config [hunk 1]
```diff
@@ -1,7 +1,7 @@
 process {
-    executor='lsf'
-    queue="$SCXA_LSF_QUEUE"
-    clusterOptions="$SCXA_LSF_OPTIONS"
+    executor='slurm'
+    queue="$SCXA_SLURM_QUEUE"
```
Comment (anilthanki) [inline nextflow.config]:
```suggestion
    queue="$SCXA_HPC_QUEUE"
```

Just so can be used in future for another scheduler if changed

File: nextflow.config [hunk 2]
```diff
@@ -28,8 +28,10 @@ env {
 
     SCXA_ENV="$SCXA_ENV"
     SCXA_WORKFLOW_ROOT="$SCXA_WORKFLOW_ROOT"
-    SCXA_LSF_QUEUE="$SCXA_LSF_QUEUE"
-    SCXA_LSF_OPTIONS="$SCXA_LSF_OPTIONS_ESCAPED"
+    #SCXA_LSF_QUEUE="$SCXA_LSF_QUEUE"
+    #SCXA_LSF_OPTIONS="$SCXA_LSF_OPTIONS_ESCAPED"
+    SCXA_SLURM_QUEUE="$SCXA_SLURM_QUEUE"
```
Comment (anilthanki) [inline nextflow.config]:
```suggestion
    SCXA_HPC_QUEUE="$SCXA_HPC_QUEUE"
```

File: nextflow.config [hunk 3]
```diff
@@ -28,8 +28,10 @@ env {
...
+    SCXA_SLURM_OPTIONS="$SCXA_SLURM_OPTIONS_ESCAPED"
```
Comment (anilthanki) [inline nextflow.config]:
```suggestion
    SCXA_HPC_OPTIONS="$SCXA_HPC_OPTIONS_ESCAPED"
```

File: nextflow.config [hunk 4]
```diff
@@ -1,7 +1,7 @@
 process {
-    executor='lsf'
-    queue="$SCXA_LSF_QUEUE"
-    clusterOptions="$SCXA_LSF_OPTIONS"
+    executor='slurm'
+    queue="$SCXA_SLURM_QUEUE"
+    clusterOptions="$SCXA_SLURM_OPTIONS"
```
Comment (anilthanki) [inline nextflow.config]:
```suggestion
    clusterOptions="$SCXA_HPC_OPTIONS"
```

File: nextflow.config (hunk 3 above)
Comment (anilthanki) [inline nextflow.config]:
```suggestion
    SCXA_HPC_OPTIONS="$SCXA_HPC_OPTIONS_ESCAPED"
```

File: nextflow.config [hunk 5]
```diff
@@ -28,8 +23,8 @@ env {
 
     SCXA_ENV="$SCXA_ENV"
     SCXA_WORKFLOW_ROOT="$SCXA_WORKFLOW_ROOT"
-    SCXA_LSF_QUEUE="$SCXA_LSF_QUEUE"
-    SCXA_LSF_OPTIONS="$SCXA_LSF_OPTIONS_ESCAPED"
+    SCXA_HPC_QUEUE="$SCXA_HPC_QUEUE"
```
Comment (anilthanki) [inline nextflow.config:25]: Just wondering where is this defined?

Comment (pmb59) [reply nextflow.config:26]: In a envs file in codon

File: bin/submitControlWorkflow.sh [hunk 6]
```diff
@@ -145,9 +145,9 @@ fi
 nextflowCommand="nextflow run -N $SCXA_REPORT_EMAIL -resume $(pwd)/workflow/${workflow}/main.nf $expNamePart $skipQuantificationPart $skipAggregationPart $tertiaryWorkflowPart $skipTertiaryPart $galaxyCredentialsPart $overwritePart --enaSshUser fg_atlas_sc -work-dir $workingDir"
 echo "$nextflowCommand"
 
-# Run the LSF submission if it's not already running
+# Run the SLURM submission if it's not already running
 
-bjobs -w | grep " ${controlJobName}" > /dev/null 2>&1
+squeue -o "%.100j" | grep -w "${controlJobName}" > /dev/null 2>&1
```
Comment (irisdianauy) [inline bin/submitControlWorkflow.sh]: Here, the following is enough for very long job names, will avoid extra spaces in output:
```suggestion
squeue -o "%j" | grep -w "${controlJobName}" > /dev/null 2>&1
```

File: bin/submitControlWorkflow.sh [hunk 7]
```diff
@@ -165,13 +165,16 @@ if [ $? -ne 0 ]; then
...
     rm -rf run.out run.err .nextflow.log*  
-    bsub \
-        -J "${controlJobName}" \
-        -M 16000 -R "rusage[mem=16000]" \
-        -u $SCXA_REPORT_EMAIL \
-        -o run.out \
-        -e run.err \
-        "$nextflowCommand" 
+
+    sbatch \
```
Comment (irisdianauy) [inline bin/submitControlWorkflow.sh:170]: Needs `--time`, otherwise the sbatch command will fail.

**Review by anilthanki — APPROVED** [review]
LGTM

File: bin/submitIndependentControlWorkflows.sh [hunk 8]
```diff
@@ -56,7 +56,7 @@ else
     touch $submissionMarker
 fi
 
-currentJobs=$(bjobs -w | grep $controlJobSuffix)
+currentJobs=$(squeue -o "%.18i %.9P %.8j %.8u %.2t %.10M %.6D %R" | grep $controlJobSuffix)
```
Comment (irisdianauy) [inline bin/submitIndependentControlWorkflows.sh]: Here, you can write all `-o %all` to output all possible info, or simply remove the column size settings to ensure that you capture the entirety of any info, e.g.,
```suggestion
currentJobs=$(squeue -o "%i %P %j %u %t %M %D %R" | grep $controlJobSuffix)
```

**Review by irisdianauy — COMMENTED** [review]
Please see comments

File: bin/submitControlWorkflow.sh (hunk 7 above)
Comment (pmb59) [reply bin/submitControlWorkflow.sh:170]: done here [46aaf0a](https://github.com/ebi-gene-expression-group/scxa-control-workflow/pull/45/commits/46aaf0a8b912ed34aa454f4bfe7825d2322800c6)

**Comment by pmb59** [comment]
Thanks @irisdianauy and @anilthanki for reviewing!

**Review by irisdianauy — APPROVED** [review]
LGTM

**Pull request merged**